sklearn/tree/_classes.py

```diff
@@ -225,46 +230,64 @@ def fit(self, X, y, sample_weight=None, check_input=True):
             y = np.ascontiguousarray(y, dtype=DOUBLE)
 
         # Check parameters
+        if self.max_depth is not None:
+            check_scalar(
+                self.max_depth,
+                name="max_depth",
+                target_type=numbers.Integral,
+                min_val=1,
+            )
         max_depth = np.iinfo(np.int32).max if self.max_depth is None else self.max_depth
-        max_leaf_nodes = -1 if self.max_leaf_nodes is None else self.max_leaf_nodes
 
         if isinstance(self.min_samples_leaf, numbers.Integral):
-            if not 1 <= self.min_samples_leaf:
-                raise ValueError(
-                    "min_samples_leaf must be at least 1 or in (0, 0.5], got %s"
-                    % self.min_samples_leaf
-                )
+            check_scalar(
+                self.min_samples_leaf,
+                name="min_samples_leaf",
+                target_type=numbers.Integral,
+                min_val=1,
+            )
             min_samples_leaf = self.min_samples_leaf
         else:  # float
-            if not 0.0 < self.min_samples_leaf <= 0.5:
-                raise ValueError(
-                    "min_samples_leaf must be at least 1 or in (0, 0.5], got %s"
-                    % self.min_samples_leaf
-                )
+            check_scalar(
+                self.min_samples_leaf,
+                name="min_samples_leaf",
+                target_type=numbers.Real,
+                min_val=0.0,
+                max_val=0.5,
```

Comment:
I think that for consistency, we need to remove this upper-bound because the integer parametrization allows for more than `int(0.5 * n_samples)`.

```suggestion
```